### PR TITLE
[IMP] base: allow loading the SMTP CLI configuration in a mail server

### PR DESCRIPTION
**Pull request by std-odoo**
Purpose
=======
On our saas, when users configure a mail server, they can't use anymore the "odoo.com" configuration (which is stored in the odoo-bin argument).

To allow them to continue using the SMTP CLI configuration, we add a new "smtp_authentication". When this authentication method is chosen, all the "connection" fields of the mail server are ignored, and the connection information are taken from the odoo-bin arguments. So they can choose the from filter of this SMTP configuration, the priority...

Task-3061882

### CODE REVIEW

File: odoo/addons/base/models/ir_mail_server.py [hunk 1]
```diff
@@ -137,6 +137,15 @@ def _compute_smtp_authentication_info(self):
                 server.smtp_authentication_info = _(
                     'Authenticate by using SSL certificates, belonging to your domain name. \n'
                     'SSL certificates allow you to authenticate your mail server for the entire domain name.')
+            elif server.smtp_authentication == 'cli':
+                smtp_server = tools.config.get('smtp_server')
+                smtp_port = tools.config.get('smtp_port', 25)
+                smtp_user = tools.config.get('smtp_user')
+
+                server.smtp_authentication_info = _(
+                    'Take the SMTP configuration from the CLI arguments. \n'
```
Comment (tde-banana-odoo) [inline odoo/addons/base/models/ir_mail_server.py:146]: Would use named parameters :)

File: odoo/addons/base/tests/test_ir_mail_server.py [hunk 2]
```diff
@@ -455,3 +456,60 @@ def test_mail_server_mail_default_from_filter(self):
...
+
+        self.assert_email_sent_smtp(
+            smtp_from='test@cli_example.com',
+            message_from='test@cli_example.com',
+            from_filter='cli_example.com',
+        )
+
+        # check that the CLI server take the configuration in the odoo-bin argument
+        # except the from_filter which is taken on the mail server
+        with self.mock_smtplib_connection():
```
Comment (tde-banana-odoo) [inline odoo/addons/base/tests/test_ir_mail_server.py:495]: Don't see any change with above test ?

File: odoo/addons/base/tests/test_ir_mail_server.py [hunk 3]
```diff
@@ -455,3 +456,60 @@ def test_mail_server_mail_default_from_filter(self):
...
+            'smtp_encryption': 'none',
+            'smtp_authentication': 'cli',
+            'from_filter': 'cli_example.com',
+        }])
+
+        IrMailServer = self.env['ir.mail_server']
+
+        with self.mock_smtplib_connection():
+            smtp_session = IrMailServer.connect(smtp_from='test@cli_example.com')
+            message = self._build_email(mail_from='test@cli_example.com')
```
Comment (tde-banana-odoo) [inline odoo/addons/base/tests/test_ir_mail_server.py:484]: Shoudln't we rely on mocked tools config ?

**Review by tde-banana-odoo — COMMENTED** [review]
Validation !

File: odoo/addons/base/models/ir_mail_server.py [hunk 4]
```diff
@@ -98,9 +98,9 @@ class IrMailServer(models.Model):
         "FROM Filtering",
         help='Define for which email address or domain this server can be used.\n'
              'e.g.: "notification@odoo.com" or "odoo.com"')
-    smtp_host = fields.Char(string='SMTP Server', required=True, help="Hostname or IP of SMTP server")
```
[the remark this answers is not in the document]
Comment (std-odoo) [reply odoo/addons/base/models/ir_mail_server.py:101]: ok to add it in the view :)

File: odoo/addons/base/tests/test_ir_mail_server.py (hunk 2 above)
Comment (std-odoo) [reply odoo/addons/base/tests/test_ir_mail_server.py:459]: good catch

File: odoo/addons/base/tests/test_ir_mail_server.py (hunk 3 above)
Comment (std-odoo) [reply odoo/addons/base/tests/test_ir_mail_server.py]: indeed, not necessary

**Review by std-odoo — COMMENTED** [review]
@tde-banana-odoo Hi :) Changes done :)

**Review by Julien00859 — COMMENTED** [review]
I had to reread multiple times the commit message and the task description to understand what's going on.

Let me rephrase it, you tell me if I understood it correctly.

When we send an email, it introspects the `From` header of the mail to determine what outgoing server to use, e.g. if I send a email as juc@example.com and that there is an outgoing mail server for example.com, it will send the mail using that outgoing mail server.

Now, some users rather use the mail server they set via the command line interface, i.e. via the `--smtp` option and related ones.

This new `cli` possible choice inside the `smtp_authentication` selection mark that this outgoing mail server is merely a proxy to what was configured via the cli. So anything configured on this outgoing mail server should be ignored in favor the cli.

Am I right?

**Review by Julien00859 — APPROVED** [review]
ok implementation wise

**Comment by tde-banana-odoo** [comment]
@robodoo r+
